### PR TITLE
Supports 'ensure exist' class addon in Addon-manager

### LINKED ISSUE #36411: Need to document the correct behaviors user should expect from Addon Manager v6.0.
The new Addon Manager being bumped up in #36008 has some fundamental behavioral changes comparing to the old one. This [README](https://github.com/kubernetes/kubernetes/blob/master/cluster/addons/README.md) explains how the new Addon Manager work and makes below claim:
> The add-ons are visible through the API (they can be listed using kubectl), but direct manipulation of these objects through Apiserver is discouraged because the system will bring them back to the original state.

However, the underlying consequences are not well covered in the case that users don't have access to the master node. Issues we've encountered listed below:
- Some applications may be broke by this new behavior, like #36194. The fact is that if the application depends on changing the addon configuration through the apiserver in order to update it, very likely this application will no longer work.
- A few addons that need to stay adjustable while running are also affected. Under this new behavior, they need to leave specific fields unset in the manifest files in order to make them tunable. Like [`kube-dns`](https://github.com/kubernetes/kubernetes/tree/master/cluster/addons/dns) leaves `replicas` unset for horizontal scaling and [`heapster`](https://github.com/kubernetes/kubernetes/tree/master/cluster/addons/cluster-monitoring/influxdb) leaves `resource` unset for vertical scaling.
- Those fileds left unset in the manifest files are tunable based on one precondition that the modification will not save the configuration of object in its annotation(`last-applied-configuration`). Otherwise the Addon Manager will still reconcile it due to the [three-way-merge](https://github.com/kubernetes/kubernetes/blob/master/pkg/util/strategicpatch/patch.go#L1165-L1247) strategy. For instance, modifications made by k8s go clients will work(like [horizontal autoscaler](https://github.com/kubernetes-incubator/cluster-proportional-autoscaler) and [vertical autoscaler](https://github.com/kubernetes/contrib/tree/master/addon-resizer)). But `kubectl apply`, `kubectl edit`, `kubectl create` etc. will not work because they will update the annotation.

In short, the behavior for modifying the adjustable fields of addons are chaos and definitely not recommended to users. We should really emphasize this in docs to reduce the potential of breaking more unexpected applications.

@mikedanese 

cc @bowei @bprashanth

### LINKED ISSUE #37047: addon manager needs a "create if not exist" mode
Because of how patch is implemented, we can't manage things like configmaps through addon manager.  If a user edits the configmap and then addon manager applies the config again, the `data` field gets clobbered.  You just can't have 2 owners of an object like that.

I think we have a class of addons that we want to assert absolute or near-absolute truth for, and a series that we just want to initialize and let the user own.  We only get away with this for replicas because of sane defaults.

@mikedanese @MrHohn

### LINKED ISSUE #39561: Default Storage Classes can't easily be disabled
PR https://github.com/kubernetes/kubernetes/pull/31617/files introduced default storage classes for 1.6. They are created with cluster addon which can not easily be modified by end users (requires access to the master).

We should improve experience before 1.6 goes out.

CC @kubernetes/sig-storage-misc

### PR DESCRIPTION
**Pull request by MrHohn**
Fixes #39561, fixes #37047 and fixes #36411. Depends on #40057.

This PR splits cluster addons into two categories:
- Reconcile: Addons that need to be reconciled (`kube-dns` for instance).
- EnsureExists: Addons that need to be exist but changeable (`default-storage-class`).

The behavior for the 'EnsureExists' class addon would be:
- Create it if not exist.
- Users could do any modification they want, addon-manager will not reconcile it.
- If it is deleted, addon-manager will recreate it with the given template.
- It will not be updated/clobbered during upgrade.

As Brian pointed out in [#37048/comment](https://github.com/kubernetes/kubernetes/issues/37048#issuecomment-272510835), this may not be the best solution for addon-manager. Though #39561 needs to be fixed in 1.6 and we might not have enough bandwidth to do a big surgery.

@mikedanese @thockin 

cc @kubernetes/sig-cluster-lifecycle-misc 

---

Tasks for this PR:
- [x] Supports 'ensure exist' class addon and switch to use new labels in addon-manager.
- [x] Updates READMEs regarding the new behavior of addon-manager.
- [x] Updated `test/e2e/addon_update.go` to match the new behavior.
- [x] Go through all current addons and apply the new labels on them regarding what they need.
- [x] Bump addon-manager and update its template files.

### CODE REVIEW

**Review by pires — APPROVED** [review]
lgtm

File: cluster/addons/addon-manager/kube-addons.sh
Comment (mikedanese) [inline cluster/addons/addon-manager/kube-addons.sh]: What is the error code of this?

**Comment by mikedanese** [comment]
This looks good. Can you bump addon manager version in this PR ?

**Comment by luxas** [comment]
Please a little bit with updating the tag. I have an open PR which I have rebased countless times that @ixdy said he was gonna take a look at today and maybe merge that updates this same file: https://github.com/kubernetes/kubernetes/pull/38926

File: cluster/addons/addon-manager/kube-addons.sh
Comment (MrHohn) [inline cluster/addons/addon-manager/kube-addons.sh]: For the `kubectl` command, try to create something that already exist will exit with 255. And successfully create all objects should exit with 0. Though another common case would be some creations succeed while some creations fail. Not sure what the exit code would be in this case.

I removed the line "completed successfully" and "completed with errors" for this function because it is somehow hard to tell whether it run correctly or not.

**Comment by MrHohn** [comment]
> Can you bump addon manager version in this PR ?

@mikedanese Sure, but we need to wait for #40057 get reviewed and merged. Otherwise this won't work.

File: cluster/addons/addon-manager/kube-addons.sh
Comment (thockin) [inline cluster/addons/addon-manager/kube-addons.sh]: should we just patch kubectl to have a `--ignore-already-exists` flag to create, similar to `--ignore-not-found` on delete?

File: cluster/addons/addon-manager/kube-addons.sh
Comment (thockin) [inline cluster/addons/addon-manager/kube-addons.sh]: rename this function to be clearer?

File: cluster/addons/addon-manager/kube-addons.sh
Comment (MrHohn) [inline cluster/addons/addon-manager/kube-addons.sh]: Makes sense, will try to patch that.

File: cluster/addons/addon-manager/kube-addons.sh
Comment (MrHohn) [inline cluster/addons/addon-manager/kube-addons.sh]: You mean `update_addons`? What about `reconcile_addons()`?

**Comment by thockin** [comment]
naming it to match the labels seems sane.

**Comment by MrHohn** [comment]
Updates:

As discussed offline, Addon-manager are migrating to use its own specific labels to classify addons. The new labels are:
- `addonmanager.kubernetes.io/reconcile=true`
- `addonmanager.kubernetes.io/ensure-exist=true`

Meanwhile, the old label `kubernetes.io/cluster-service` will be deprecated (only for Addon-manager). But for now, the old label will still be respected. Enumerate cases below:
- Have `kubernetes.io/cluster-service` only: being reconciled as before.
- Have `addonmanager.kubernetes.io/reconcile=true` only: being reconciled.
- Have `kubernetes.io/cluster-service` and `addonmanager.kubernetes.io/reconcile=true`: being reconciled.
- Have `kubernetes.io/cluster-service` and `addonmanager.kubernetes.io/ensure-exist=true`: ensuring it exist.
- Have `addonmanager.kubernetes.io/ensure-exist=true` only: ensuring it exist.

More detail could be found in the updated READMEs.

Tasks for this PR:
- [x] Supports 'ensure exist' class addon and switch to use new labels in Addon-manager.
- [x] Updates READMEs regarding the new behavior of Addon-manager.
- [ ] Bump Addon-manager to a v6.3 and update its template files.
- [ ] Go through all current addons and apply the new labels on them regarding what they need.
- [ ] Updated `test/e2e/addon_update.go` to match its new behavior.

Push the first two commits ahead to check it is on the right track.

File: cluster/addons/addon-manager/kube-addons.sh
Comment (MrHohn) [inline cluster/addons/addon-manager/kube-addons.sh]: These two `-l` mean:
```
if match ${ADDON_RECONCILE_LABEL}=true ||
    (match ${CLUSTER_SERVICE_LABEL}=true && don't have ${ADDON_ENSURE_EXIST_LABEL}) {
    reconcile_it()...
}
```

File: cluster/addons/addon-manager/kube-addons.sh
Comment (MrHohn) [inline cluster/addons/addon-manager/kube-addons.sh]: Also bad news on this one. I tried implement `--ignore-already-exists` for `kubectl create` but encountered another problem: `kubectl create` returns error when there is no object passed to create. With this flag enabled, if we pass in a few objects to create but all of them exist, we still get an error because the "nothing create error".

I need to put more thoughts on it. But feel like this could be done later.

**Comment by spxtr** [comment]
@k8s-bot bazel test this
#40526

**Comment by thockin** [comment]
Why 2 labels:

* addonmanager.kubernetes.io/reconcile=true
* addonmanager.kubernetes.io/ensure-exist=true

vs a single label and enumerated value?

* addonmanager.kubernetes.io/mode=Reconcile
* addonmanager.kubernetes.io/mode=EnsureExists

**Comment by MrHohn** [comment]
@thockin Yeah single label seems better. Label value=true/false does not make much sense.

File: cluster/addons/addon-manager/kube-addons.sh
Comment (MrHohn) [inline cluster/addons/addon-manager/kube-addons.sh]: Sorry, this usage for multiple selectors will not work. Discussion on #40588.

**Comment by ixdy** [comment]
fyi https://github.com/kubernetes/kubernetes/pull/38926 bumps addon-manager to v6.3, and the image has already been pushed, though the specs/salt haven't been updated yet.

File: cluster/addons/README.md
Comment (thockin) [inline cluster/addons/README.md]: s/periodical/periodic

File: cluster/addons/README.md
Comment (thockin) [inline cluster/addons/README.md]: Reconcile is capitalized.

Comment (thockin) [inline cluster/addons/README.md]: I would only quote "Reconcile" here

File: cluster/addons/README.md
Comment (thockin) [inline cluster/addons/README.md]: clobber

File: cluster/addons/addon-manager/README.md
Comment (thockin) [inline cluster/addons/addon-manager/README.md]: addon-manager to match the dir name.

File: cluster/addons/addon-manager/README.md
Comment (thockin) [inline cluster/addons/addon-manager/README.md]: "of these addons"

apiserver or "the API server"

File: cluster/addons/addon-manager/README.md
Comment (thockin) [inline cluster/addons/addon-manager/README.md]: addon-manager, no caps

File: cluster/addons/addon-manager/README.md
Comment (thockin) [inline cluster/addons/addon-manager/README.md]: will be checked for existence only.

File: cluster/addons/addon-manager/README.md
Comment (thockin) [inline cluster/addons/addon-manager/README.md]: Users

File: cluster/addons/addon-manager/README.md
Comment (thockin) [inline cluster/addons/addon-manager/README.md]: s/such resource exist/instance of the resource with that name/

File: cluster/addons/addon-manager/README.md
Comment (thockin) [inline cluster/addons/addon-manager/README.md]: Deprecation policy says 1 year.

File: cluster/addons/addon-manager/README.md
Comment (thockin) [inline cluster/addons/addon-manager/README.md]: Resources

Comment (thockin) [inline cluster/addons/addon-manager/README.md]: s/needs/need

File: cluster/addons/addon-manager/README.md
Comment (thockin) [inline cluster/addons/addon-manager/README.md]: resources need

Comment (thockin) [inline cluster/addons/addon-manager/README.md]: Otherwise they

File: cluster/addons/addon-manager/kube-addons.sh
Comment (thockin) [inline cluster/addons/addon-manager/kube-addons.sh]: ping me when you change this to one label (if that is still happening)

**Review by thockin — COMMENTED** [review]
Mostly just English nits.  waiting for annotation format change

**Comment by MrHohn** [comment]
Sorry for the confusing state here. Changed to use one label with modes. But this implementation is now buggy after I split the `kubectl apply -l XXX -l XXX`, which actually won't work, into two commands. One related issue is #40635.

Will update with more details and cases later.

**Comment by MrHohn** [comment]
@k8s-bot cross build this

**Comment by MrHohn** [comment]
Seems like all the puzzles have been solved. The new commits update addon manager e2e test and add  "addonmanager.kubernetes.io/mode" label to cluster addons. Right now the only "ensure exist" class addon is `default-storage-class`.

This should also fix #39561. cc @msau42

Ready for review now. @mikedanese  @thockin

File: cluster/saltbase/salt/kube-addons/kube-addon-manager.yaml
Comment (MrHohn) [inline cluster/saltbase/salt/kube-addons/kube-addon-manager.yaml]: Will push the new images after approved.

**Comment by thockin** [comment]
I glanced.  It looks good overall.  I abdicate any authority I had on final review to @mikedanese :)

**Comment by msau42** [comment]
cc @jsafrane

**Comment by MrHohn** [comment]
Friendly ping since it is getting close to code freeze :) @mikedanese 

(The conflicts are all version tweaks and unrelated changes so I kept them there.)

**Comment by thockin** [comment]
@mikedanese might be busy this week - can we nominate an alternate reviewer?

**Comment by MrHohn** [comment]
@luxas @pires @jbeda Would anyone has bandwidth to help on the final review? Sorry for getting you involved, but we really need this in for the default storage class fix :(

**Comment by roberthbailey** [comment]
@zmerlynn would be a good reviewer candidate if he has time to take a look.

**Comment by pires** [comment]
Unfortunately, don't have the bw to review the changes before code-freeze. It's been a while since I last looked into this.

**Comment by msau42** [comment]
Can someone please review this change?  We really need this in 1.6 to allow users to disable default storage classes.

**Comment by thockin** [comment]
This LGTM.  Needs rebase

**Comment by MrHohn** [comment]
Many thanks for taking the time! Rebased and pushed the real images:
- gcr.io/google-containers/kube-addon-manager:v6.4-alpha.3
- gcr.io/google-containers/kube-addon-manager-amd64:v6.4-alpha.3
- gcr.io/google-containers/kube-addon-manager-arm:v6.4-alpha.3
- gcr.io/google-containers/kube-addon-manager-arm64:v6.4-alpha.3
- gcr.io/google-containers/kube-addon-manager-ppc64le:v6.4-alpha.3
- gcr.io/google-containers/kube-addon-manager-s390x:v6.4-alpha.3

Also verified the slow test passed locally :)

**Comment by MrHohn** [comment]
BTW, also worth to mention that the new images are built with kubectl binary on HEAD (17375fc59fff39135af63bd1750bb07c36ef873b) instead of the released one to unblock #41765.

**Pull request merged**
